Mathlib/FieldTheory/IsSepClosed.lean

```diff
@@ -0,0 +1,174 @@
+/-
+Copyright (c) 2020 Kenny Lau, 2023 Jz Pan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau, Jz Pan
+-/
+import Mathlib.FieldTheory.IsAlgClosed.AlgebraicClosure
+
+/-!
+# Separably Closed Field
+
+In this file we define the typeclass for separably closed fields and separable closures,
+and prove some of their properties.
+
+## Main Definitions
+
+- `IsSepClosed k` is the typeclass saying `k` is a separably closed field, i.e. every separable
+polynomial in `k` splits.
+
+- `IsSepClosure R K` is the typeclass saying `K` is a separable closure of `R`, where `R` is a
+  commutative ring. This means that the map from `R` to `K` is injective, and `K` is
+  separably closed and separable over `R`
+
+## Tags
+
+separable closure, separably closed
+
+-/
+
+universe u v w
+
+open scoped Classical BigOperators Polynomial
+
+open Polynomial
+
+variable (k : Type u) [Field k]
+
+/-- Typeclass for separably closed fields.
+
+To show `Polynomial.Splits p f` for an arbitrary ring homomorphism `f`,
+see `IsSepClosed.splits_codomain` and `IsSepClosed.splits_domain`.
+-/
+class IsSepClosed : Prop where
+  splits_of_separable : ∀ p : k[X], p.Separable → (p.Splits <| RingHom.id k)
+
+/-- Every separable polynomial splits in the field extension `f : K →+* k` if `k` is
+separably closed.
+
+See also `IsSepClosed.splits_domain` for the case where `K` is separably closed.
+-/
+theorem IsSepClosed.splits_codomain {k K : Type _} [Field k] [IsSepClosed k] [Field K] {f : K →+* k}
+    (p : K[X]) (h : p.Separable) : p.Splits f := by
+  convert IsSepClosed.splits_of_separable (p.map f) (Separable.map h); simp [splits_map_iff]
+
+/-- Every separable polynomial splits in the field extension `f : K →+* k` if `K` is
+separably closed.
+
+See also `IsSepClosed.splits_codomain` for the case where `k` is separably closed.
+-/
+theorem IsSepClosed.splits_domain {k K : Type _} [Field k] [IsSepClosed k] [Field K] {f : k →+* K}
+    (p : k[X]) (h : p.Separable) : p.Splits f :=
+  Polynomial.splits_of_splits_id _ <| IsSepClosed.splits_of_separable _ h
```

Comment:
I personally find reading these two lemmas quite confusing because of the change in the type of `f`. In the first lemma you have `f : K →+* k` and in the second you have `f : k →+* K`. I think that nowadays in mathlib it's traditional to have one "style" for notation throughout a file (e.g. "K is always bigger than k") and then to stick to this style throughout the file. Otherwise the file can become quite confusing to read. I've suggested a change to the first lemma which I think sticks to this style.